src/transport.rs

```diff
@@ -121,107 +112,68 @@ impl Transport {
+                let status = res.status();
+                match status {
+                    StatusCode::OK
+                    | StatusCode::CREATED
+                    | StatusCode::SWITCHING_PROTOCOLS
+                    | StatusCode::NO_CONTENT => future::ok(res),
+                    _ => future::err(Error::Fault {
+                        code: status,
+                        // TODO get_error_message
+                        message: status
+                            .canonical_reason()
+                            .unwrap_or("unknown error code")
+                            .to_owned(),
+                    }),
+                }
+            })
+            .map(|r| {
+                r.into_body()
+                    .map(|chunk| chunk.into_iter().collect::<Vec<u8>>())
+                    .map_err(Error::Hyper)
+            })
+            .flatten_stream()
     }
 
     fn send_request(
         &self,
         req: Request<hyper::Body>,
-    ) -> Result<hyper::Response<Body>> {
+    ) -> impl Future<Item = hyper::Response<Body>, Error = Error> {
         let req = match self {
             Transport::Tcp { ref client, .. } => client.request(req),
             Transport::EncryptedTcp { ref client, .. } => client.request(req),
             #[cfg(feature = "unix-socket")]
             Transport::Unix { ref client, .. } => client.request(req),
         };
 
-        self.runtime().block_on(req).map_err(Error::Hyper)
-    }
-
-    fn runtime(&self) -> RefMut<Runtime> {
-        match self {
-            Transport::Tcp { ref runtime, .. } => runtime.borrow_mut(),
-            Transport::EncryptedTcp { ref runtime, .. } => runtime.borrow_mut(),
-            #[cfg(feature = "unix-socket")]
-            Transport::Unix { ref runtime, .. } => runtime.borrow_mut(),
-        }
+        req.map_err(Error::Hyper)
     }
 
-    /// Extract the error message content from an HTTP response that
-    /// contains a Docker JSON error structure.
-    fn get_error_message(
-        &self,
-        res: Response<Body>,
-    ) -> Option<String> {
-        let chunk = match self.runtime().block_on(res.into_body().concat2()) {
-            Ok(c) => c,
-            Err(..) => return None,
-        };
-
-        match String::from_utf8(chunk.into_iter().collect()) {
-            Ok(output) => {
-                let json_response = serde_json::from_str::<Value>(output.as_str()).ok();
-                json_response
-                    .as_ref()
-                    .and_then(|x| x.as_object())
-                    .and_then(|x| x.get("message"))
-                    .and_then(|x| x.as_str())
-                    .map(|x| x.to_owned())
-            }
-            Err(..) => None,
-        }
-    }
+    // Extract the error message content from an HTTP response that
+    // contains a Docker JSON error structure.
+    // fn get_error_message(&self, body: &str) -> Option<String> {
+    //     match String::from_utf8(chunk.into_iter().collect()) {
```

Comment:
now that we're using serde, if there's a standard way to represent it with a struct. we have more civilized tools now!

Reply:
Yeah, I'm struggling a bit as to how best to reintroduce this method, because in the error case we need to read the whole body and parse it to get the error message, but in the success page we just want to return a stream of body chunks... Haven't find the right sequence of combinators yet :)